com-api/examples/basic-consumer-producer.rs

```diff
@@ -38,40 +38,49 @@ impl<R: Runtime> VehicleMonitor<R> {
             Err(e) => Err(e),
         }
     }
-}
 
+    pub fn write_tire_data(&self, tire: Tire) -> Result<()> {
+        let uninit_sample = self.producer.left_tire.allocate()?;
+        let sample = uninit_sample.write(tire);
+        sample.send()?;
+        Ok(())
+    }
+}
 
-fn use_consumer<R: Runtime>(runtime: &R)
-{
-    // Create service discovery
-    let consumer_discovery = runtime.find_service::<VehicleInterface>(InstanceSpecifier::new("My/Funk/ServiceName").unwrap());
+fn use_consumer<R: Runtime>(runtime: &R) -> VehicleConsumer<R> {
+    // Find all the avaiable service instances using ANY specifier
+    let consumer_discovery = runtime.find_service::<VehicleInterface>(FindServiceSpecifier::Any);
     let available_service_instances = consumer_discovery.get_available_instances().unwrap();
 
-    // Create consumer from first discovered service
     let consumer_builder = available_service_instances
         .into_iter()
-        .find(|desc| desc.get_instance_id() == 42)
+        .find(|desc| desc.get_instance_identifier() == "My/Funk/ServiceName")
         .unwrap();
+
     let consumer = consumer_builder.build().unwrap();
-    VehicleMonitor::new(consumer).monitor_tire_data().unwrap();
+    //
+    consumer
 }
 
-fn use_producer<R: Runtime>(runtime: &R)
-{
-    let producer_builder = runtime.producer_builder::<VehicleInterface, VehicleProducer<R>>(InstanceSpecifier::new("My/Funk/ServiceName").unwrap());
+fn use_producer<R: Runtime>(runtime: &R) -> VehicleOfferedProducer<R> {
+    let producer_builder = runtime.producer_builder::<VehicleInterface, VehicleProducer<R>>(
+        InstanceSpecifier::new("My/Funk/ServiceName").unwrap(),
+    );
     let producer = producer_builder.build().unwrap();
     let offered_producer = producer.offer().unwrap();
-
-    // Business logic
-    let uninit_sample = offered_producer.left_tire.allocate().unwrap();
-    let sample = uninit_sample.write(Tire {});
-    sample.send().unwrap();
+    offered_producer
 }
 
 fn run_with_runtime<R: Runtime>(name: &str, runtime: &R) {
     println!("\n=== Running with {} runtime ===", name);
-    use_producer(runtime);
-    use_consumer(runtime);
+
+    let monitor = VehicleMonitor::new(use_consumer(runtime), use_producer(runtime));
```

Comment:
```suggestion
    let producer = use_producer(runtime);
    let consumer = use_consumer(runtime);
    let monitor = VehicleMonitor::new(consumer, producer);
```
How about making producer first to make sure consumer will find suitable one?

Reply:
yes in this example we can move it., but both should be independent of each other.

Reply:
Thats not correct I think. 
We can't just create Consumer, we can only create Consumer for previously discovered Producer so correct order of operation is mandatory.